mesonbuild/modules/unstable_external_project.py

```diff
@@ -0,0 +1,250 @@
+        if link_exelist:
+            self.run_env['LD'] = self._quote_and_join(link_exelist)
+        self.run_env['LDFLAGS'] = self._quote_and_join(link_args)
+        self.run_env = self.user_env.get_env(self.run_env)
+        PkgConfigDependency.setup_env(self.run_env, self.env, MachineChoice.HOST,
+                                      os.path.join(self.env.get_build_dir(), 'meson-uninstalled'))
+        self._run('configure', configure_cmd)
+
+    def _quote_and_join(self, array):
+        return ' '.join([shlex.quote(i) for i in array])
+
+    def _validate_configure_options(self, required_keys):
+        # Ensure the user at least try to pass basic info to the build system,
+        # like the prefix, libdir, etc.
+        for key in required_keys:
+            key_format = '@{}@'.format(key)
+            for option in self.configure_options:
+                if key_format in option:
+                    break
+            else:
+                m = 'At least one configure option must contain "{}" key'
+                raise InterpreterException(m.format(key_format))
+
+    def _format_options(self, options, variables):
+        out = []
+        missing = set()
+        regex = get_variable_regex('meson')
+        confdata = {k: (v, None) for k, v in variables.items()}
+        for o in options:
+            arg, missing_vars = do_replacement(regex, o, 'meson', confdata)
+            missing.update(missing_vars)
+            out.append(arg)
+        if missing:
+            var_list = ", ".join(map(repr, sorted(missing)))
+            raise EnvironmentException(
+                "Variables {} in configure options are missing.".format(var_list))
+        return out
+
+    def _run(self, step, command):
+        mlog.log('External project {}:'.format(self.name), mlog.bold(step))
+        output = None if self.verbose else subprocess.DEVNULL
+        p, o, e = Popen_safe(command, cwd=self.build_dir, env=self.run_env,
+                                      stderr=subprocess.STDOUT,
+                                      stdout=output)
+        if p.returncode != 0:
+            m = '{} step failed:\n{}'.format(step, e)
+            raise MesonException(m)
+
+    def _get_targets(self):
+        cmd = self.env.get_build_command()
+        cmd += ['--internal', 'externalproject',
+                '--name', self.name,
+                '--srcdir', self.src_dir,
+                '--builddir', self.build_dir,
+                '--installdir', self.install_dir,
+                '--make', self.make,
+                ]
+        if self.verbose:
+            cmd.append('--verbose')
+
+        target_kwargs = {'output': '{}.stamp'.format(self.name),
+                         'depfile': '{}.d'.format(self.name),
+                         'command': cmd + ['@OUTPUT@', '@DEPFILE@'],
+                         'console': True,
+                         }
+        self.target = build.CustomTarget(self.name,
```

Comment:
`self.target` is not defined in `__init__`.

Reply:
Is that a problem? I could maybe call _get_targets() from __init__(), is that better?

Reply:
It's mostly to so that all `self.*` are declared in `__init__()`. Otherwsie, accessing them before they are set in another function will result in an error. Also, IDEs will be able to do better autocompletion :)

I was thinking of setting `self.target = None` and then checking (or asserting) that `self.target` was set when it is used.

Reply:
ok, done.